[SMTChecker] Tutorial and docs cleanup (#11075)
ethereum/solidity: docs/smtchecker.rst

below.

.. code-block:: Solidity

    pragma experimental SMTChecker;

    contract Robot {
        int x = 0;
        int y = 0;

        modifier wall {
            require(x > type(int128).min && x < type(int128).max);
            require(y > type(int128).min && y < type(int128).max);
            _;
        }

        function moveLeftUp() wall public {
            --x;
            ++y;
        }

        function moveLeftDown() wall public {
            --x;
            --y;
        }

        function moveRightUp() wall public {
            ++x;
            ++y;
        }

        function moveRightDown() wall public {
            ++x;
            --y;
        }

        function inv() public view {
            assert((x + y) % 2 == 0);
        }
    }

Function ``inv`` represents an invariant of the state machine that ``x + y``
must be even.
The SMTChecker manages to prove that regardless how many commands we give the
robot, even if infinitely many, the invariant can *never* fail. The interested
reader may want to prove that fact manually as well.  Hint: this invariant is
inductive.

We can also trick the SMTChecker into giving us a path to a certain position.

> We know that position (2, 4) is reachable, because 2 + 4 = 6 is even. Let us add

Reviewer: You proved that all reachable positions are even, but here you are invoking the converse (that all even positions are reachable). This happens to be true but does not follow

Reviewer: I'm not sure the converse is expressible in the SMTChecker right now actually... It would need to be some form of existential quantification I think

Author: Ah, that's true! I can change the text then, formulating exactly like you put, and that we want to ask whether a certain position is reachable.

the property that (2, 4) is *not* reachable, by adding the following function.

.. code-block:: Solidity

    function reach_2_4() public view {
        assert(!(x == 2 && y == 4));
    }

We know that this property is false, but while proving that the property is false,
the SMTChecker gives us exactly *how* to reach (2, 4):

.. code-block:: bash

  Warning: CHC: Assertion violation happens here.
  Counterexample:
  x = 2, y = 4
  
  Transaction trace:
  Robot.constructor()
  State: x = 0, y = 0
  Robot.moveLeftUp()
  State: x = (- 1), y = 1
  Robot.moveRightUp()
  State: x = 0, y = 2
  Robot.moveRightUp()
  State: x = 1, y = 3
  Robot.moveRightUp()
  State: x = 2, y = 4
  Robot.reach_2_4()
    --> r.sol:35:4:
     |
  35 | 			assert(!(x == 2 && y == 4));
     | 			^^^^^^^^^^^^^^^^^^^^^^^^^^^

External Calls and Reentrancy
=============================

Every external call is treated as a call to unknown code by the SMTChecker.

> The reasoning behind that is, that even if the code of the called contract is

Reviewer: ```suggestion
The reasoning behind that is that even if the code of the called contract is
```

The solver is able to infer that when ``unknown.run()`` is called, the contract
is already "locked", so it would not be possible to change the value of ``x``,
regardless what the unknown called code does.

If we "forget" to use the ``mutex`` modifier on function ``set``, the
SMTChecker is able to synthesize the behavior of the externally called code so
that the assertion fails:

.. code-block:: bash

  Warning: CHC: Assertion violation happens here.
  Counterexample:
  x = 1, lock = true, unknown = 1
  
  Transaction trace:
  Mutex.constructor(1)
  State: x = 0, lock = false, unknown = 1
  Mutex.run()
      unknown.run() -- untrusted external call, synthesized as:
          Mutex.set(1) -- reentrant call
    --> m.sol:32:3:
     |
  32 | 		assert(xPre == x);
     | 		^^^^^^^^^^^^^^^^^

.. _smtchecker_options:

*****************************
SMTChecker Options and Tuning
*****************************

Timeout
=======

The SMTChecker uses a hardcoded resource limit (``rlimit``) chosen per solver,
which is not precisely related to time. We chose the ``rlimit`` option as the default
because it gives more determinism guarantees than time inside the solver.

> This options translates roughly to "a few seconds timeout". Of course many properties

Reviewer: per query?

.. _formal_verification:

##################################
SMTChecker and Formal Verification
##################################

> Using formal verification, it is possible to perform an automated mathematical

Reviewer 2: ```suggestion
Using formal verification it is possible to perform an automated mathematical
```

proof that your source code fulfills a certain formal specification.
The specification is still formal (just as the source code), but usually much
simpler.

Note that formal verification itself can only help you understand the
difference between what you did (the specification) and how you did it
(the actual implementation). You still need to check whether the specification
is what you wanted and that you did not miss any unintended effects of it.

> Solidity implements a formal verification approach based on SMT and Horn solving.

Reviewer 2: Could link to wikipedia (or some other explanation) for "smt" and "horn solving".

The SMTChecker module automatically tries to prove that the code satisfies the

> specification given by ``require/assert`` statements. That is, it considers

Reviewer 2: ```suggestion
specification given by ``require`` and ``assert`` statements. That is, it considers
```

``require`` statements as assumptions and tries to prove that the conditions
inside ``assert`` statements are always true.  If an assertion failure is

> found, a counterexample may be given to the user, showing how the assertion can

Reviewer 2: ```suggestion
found, a counterexample may be given to the user showing how the assertion can
```

be violated. If no warning is given by the SMTChecker for a property,
it means that the property is safe.

The other verification targets that the SMTChecker checks at compile time are:

- Arithmetic underflow and overflow.
- Division by zero.
- Trivial conditions and unreachable code.
- Popping an empty array.
- Insufficient funds for a transfer.

The potential warnings that the SMTChecker reports are:

- ``<failing  property> happens here.``. This means that the SMTChecker proved that a certain property fails. A counterexample may be given, however in complex situations it may also not show a counterexample. This result may also be a false positive in certain cases, when the SMT encoding adds abstractions for Solidity code that is either hard or impossible to express.

> - ``<failing property> might happen here``. This means that the solver could not prove either case in the given timeout. Since the result is unknown, the SMTChecker reports the potential failure for soundness. This may be solved by increasing the query timeout, but the problem might also simply be too hard.

Reviewer 2: ```suggestion
- ``<failing property> might happen here``. This means that the solver could not prove either case within the given timeout. Since the result is unknown, the SMTChecker reports the potential failure for soundness. This may be solved by increasing the query timeout, but the problem might also simply be too hard for the engine to solve.
```

It is currently an experimental feature, therefore in order to use it you need
to enable it via :ref:`a pragma directive<smt_checker>`.

.. note::
  The lack of warnings for a verification target represents an undisputed
  mathematical proof of correctness, assuming no bugs in the SMTChecker and

>   underlying solver. Keep in mind that these problems are

Reviewer 2: ```suggestion
  the underlying solver. Keep in mind that these problems are
```

  *very hard* and sometimes *impossible* to solve automatically in the
  general case.  Therefore, several properties might not be solved or might
  lead to false positives for large contracts. Every proven property should
  be seen as an important achievement. For advanced users, see :ref:`SMTChecker Tuning <smtchecker_options>`
  to learn a few options that might help proving more complex
  properties.

********
Tutorial
********

Overflow
========

.. code-block:: Solidity

    pragma solidity >=0.8.0;
    pragma experimental SMTChecker;

>     // This may report a warning if no SMT solver available.

Reviewer 2: ```suggestion
    // This may report a warning if no SMT solver is available.
```

            _;
        }

        function moveLeftUp() wall public {
            --x;
            ++y;
        }

        function moveLeftDown() wall public {
            --x;
            --y;
        }

        function moveRightUp() wall public {
            ++x;
            ++y;
        }

        function moveRightDown() wall public {
            ++x;
            --y;
        }

        function inv() public view {
            assert((x + y) % 2 == 0);
        }
    }

Function ``inv`` represents an invariant of the state machine that ``x + y``
must be even.
The SMTChecker manages to prove that regardless how many commands we give the
robot, even if infinitely many, the invariant can *never* fail. The interested
reader may want to prove that fact manually as well.  Hint: this invariant is
inductive.

We can also trick the SMTChecker into giving us a path to a certain position we
think might be reachable.  We can add the property that (2, 4) is *not*
reachable, by adding the following function.

.. code-block:: Solidity

    function reach_2_4() public view {
        assert(!(x == 2 && y == 4));
    }

This property is false, and while proving that the property is false,

> the SMTChecker tells us exactly *how* to reach (2, 4):

Reviewer 2: Maybe you want to explain that this trace is not exactly deterministic (it shows *a potential path*). It could show different paths on different machines, different (solidity,z3,chc,..) releases, different resource limits.

compute ``ecrecover``, but by modelling the function calls as uninterpreted
functions we know that the return value is the same when called on equivalent
parameters. This is enough to prove that the assertion above is always true.

Abstracting a function call with an UF can be done for functions known to be
deterministic, and can be easily done for pure functions.  It is however
difficult to do this with general external functions, since they might depend
on state variables.

Reference Types and Aliasing
============================

Solidity implements aliasing for reference types with the same :ref:`data
location<data-location>`.
That means one variable may be modified through a reference to the same data
area.
The SMTChecker does not keep track of which references refer to the same data.
This implies that whenever a local reference or state variable of reference
type is assigned, all knowledge regarding variables of the same type and data
location is erased.
If the type is nested, the knowledge removal also includes all the prefix base
types.

::

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity >=0.8.0;

>     pragma abicoder v2;

Reviewer 2: Does example depend on this? Also it is default n 0.8.0

Author: It depends on v2 because of `uint[][]` parameter in a public function. But true it's default now, so will remove.

            d[0] = 2;
            // Fails as false positive because of the assignment above.
            assert(array1[0] == 42);
            // Fails because `a == b` is possible.
            assert(a[0] == 2);
            // Fails because `c[i] == b` is possible.
            assert(c[0][0] == 2);
            assert(d[0] == 2);
            assert(b[0] == 1);
        }
        function g(
            uint[] memory a,
            uint[] memory b,
            uint[][] memory c,
            uint x
        ) public {
            f(a, b, c, array2[x]);
        }
    }

After the assignment to ``b[0]``, we need to clear knowledge about ``a`` since
it has the same type (``uint[]``) and data location (memory).  We also need to
clear knowledge about ``c``, since its base type is also a ``uint[]`` located
in memory. This implies that some ``c[i]`` could refer to the same data as
``b`` or ``a``.

Notice that we do not clear knowledge about ``array`` and ``d`` because they
are located in storage, even though they also have type ``uint[]``.  However,
if ``d`` was assigned, we would need to clear knowledge about ``array`` and
vice-versa.

**********************

> Real World Assumptions

Reviewer 2: Is this section ready? What is the actual (upper bound) assumption regarding arrays in the SMTChecker?

Author: This section was just copied from before, and wasn't changed at all by this PR. The assumption is that an array's length is always < type(uint256).max

be violated. If no warning is given by the SMTChecker for a property,
it means that the property is safe.

The other verification targets that the SMTChecker checks at compile time are:

- Arithmetic underflow and overflow.
- Division by zero.
- Trivial conditions and unreachable code.
- Popping an empty array.
- Insufficient funds for a transfer.

The potential warnings that the SMTChecker reports are:

- ``<failing  property> happens here.``. This means that the SMTChecker proved that a certain property fails. A counterexample may be given, however in complex situations it may also not show a counterexample. This result may also be a false positive in certain cases, when the SMT encoding adds abstractions for Solidity code that is either hard or impossible to express.
- ``<failing property> might happen here``. This means that the solver could not prove either case within the given timeout. Since the result is unknown, the SMTChecker reports the potential failure for soundness. This may be solved by increasing the query timeout, but the problem might also simply be too hard for the engine to solve.

It is currently an experimental feature, therefore in order to use it you need

> to enable it via :ref:`a pragma directive<smt_checker>`.

Reviewer 3: Maybe this should be the first thing to mention, so maybe should be moved to top? But instead of `It` in that case you can use `Formal verification ...` or something similar. But, that is just my feeling, maybe it is fine as is...

Author: @mijovic I would prefer to leave it there because we first explain what it does, and then how to use it. I can still change it if you strongly feel like it should come before.

Reviewer 3: Let's leave it like it is, I didn't have strong opinion on that, just felt like it could be moved. But, since it is not a huge paragraph before I think it is completely fine as is.

be violated. If no warning is given by the SMTChecker for a property,
it means that the property is safe.

The other verification targets that the SMTChecker checks at compile time are:

- Arithmetic underflow and overflow.
- Division by zero.
- Trivial conditions and unreachable code.
- Popping an empty array.
- Insufficient funds for a transfer.

The potential warnings that the SMTChecker reports are:

- ``<failing  property> happens here.``. This means that the SMTChecker proved that a certain property fails. A counterexample may be given, however in complex situations it may also not show a counterexample. This result may also be a false positive in certain cases, when the SMT encoding adds abstractions for Solidity code that is either hard or impossible to express.
- ``<failing property> might happen here``. This means that the solver could not prove either case in the given timeout. Since the result is unknown, the SMTChecker reports the potential failure for soundness. This may be solved by increasing the query timeout, but the problem might also simply be too hard.

.. note::
  The lack of warnings for a verification target represents an undisputed

>   mathematical proof of correctness. Keep in mind that these problems are

Reviewer 4: This is a very strong statement. Should we add that "provided there are no bugs in the encoding or solver"?

Author: Well, I thought that was implicit haha
It's also only true if the specification is correct. So there are a few implicit assumptions to that statement. But I can add that, yes.

  *very hard* and sometimes *impossible* to solve automatically in the
  general case.  Therefore, several properties might not be solved or might
  lead to false positives for large contracts. Every proven property should
  be seen as an important achievement. For advanced users, see `ref: ...`
  to learn a few SMTChecker options that might help proving more complex
  properties.
  
It is currently an experimental feature, therefore in order to use it you need
to enable it via :ref:`a pragma directive<smt_checker>`.

***************
Simple Tutorial
***************

.. code-block:: Solidity

>     pragma experimental SMTChecker;

Reviewer 4: Missing `pragma solidity version`

    Counterexample:
    x = 1, y = 115792089237316195423570985008687907853269984665640564039457584007913129639935
     = 0
    
    Transaction trace:
    Overflow.constructor(1, 115792089237316195423570985008687907853269984665640564039457584007913129639935)
    State: x = 1, y = 115792089237316195423570985008687907853269984665640564039457584007913129639935
    Overflow.stateAdd()
        Overflow.add(1, 115792089237316195423570985008687907853269984665640564039457584007913129639935) -- internal call
     --> o.sol:9:20:
      |
    9 |             return _x + _y;
      |                    ^^^^^^^

If we add ``require`` statements that filter out overflow cases,
the SMTChecker proves that no overflow is reachable:

.. code-block:: Solidity

    pragma experimental SMTChecker;

    contract Overflow {
        uint immutable x;
        uint immutable y;

        function add(uint _x, uint _y) internal pure returns (uint) {
            return _x + _y;
        }

        constructor(uint _x, uint _y) {
            (x, y) = (_x, _y);
        }

        function stateAdd() public view returns (uint) {
            require(x < type(uint128).max);

>             require(y < type(uint128).max);

Reviewer 4: `require(x < type(uint248).max)` and `require(y < type(uint248).max)` is a more fun example. Fine with keeping this.

Author: Yea I thought about that too, but felt like 128 is more "relatable" or more common. I'm also fine with either.

    }

Assert
======

An assertion represents an invariant in your code: a property that must be true

> *for all cases*, otherwise there is a bug.

Reviewer 5: must be true *at all times* or *in all cases*? But fine with keeping it, though I'd prefer *at all times*.

Author: hmmm depends what time means. That property may be broken 2 lines earlier or two lines after, but it's true at that line

Reviewer 5: "for all cases" sounds rather non-idiomatic to me in any case...
"must be true every time the assertion is executed" or "must always be true at the point of the assertion" or...
But feel free to dismiss :-).

Author: I feel like that doesn't bring the point I'm trying to make, which is that it's true for all input and storage values

Reviewer 5: Then make it "a property that must be true for all input and storage values" maybe :-)?

Reviewer 5: But yeah, all fine with me.

Author: Changed to something different but still related.

.. code-block:: Solidity

    pragma experimental SMTChecker;

    contract Monotonic {
        function f(uint _x) internal pure returns (uint) {
            require(_x < type(uint128).max);
            return _x * 42;
        }

        function inv(uint _a, uint _b) public pure {
            require(_b > _a);
            assert(f(_b) > f(_a));
        }
    }

We can also add assertions inside loops to verify more complicated properties.
The following code searches for the maximum element of an unrestricted array of
numbers, and asserts the property that the found element must be greater or
equal every element in the array.

.. code-block:: Solidity

  pragma experimental SMTChecker;

  contract Max {
      function max(uint[] memory _a) public pure returns (uint) {
          uint m = 0;
          for (uint i = 0; i < _a.length; ++i)
              if (_a[i] > m)
                  m = _a[i];

          for (uint i = 0; i < _a.length; ++i)
              assert(m >= _a[i]);

          return m;
      }
  }

Note that in this example the SMTChecker will automatically try to prove three properties:

1. ``++i`` at line 6 does not overflow.
2. ``++i`` at line 10 does not overflow.

> 3. Assertion at line 11 is always true.

Reviewer 5: Maybe say "++i in the first loop" and "++i in the second loop" and just "the assertion" (there is only one)? Then I don't have to count lines :-)...

State Properties
================

So far the examples only demonstrated the use of the SMTChecker over pure code,
proving properties about specific operations or algorithms.
A common type of properties in smart contracts are properties that involve the
state of the contract. Multiple transactions might be needed to make an assertion
fail for such a property.

As an example, consider a 2D grid where both axis have coordinates in the range (-2^128, 2^128 - 1).

> Let us place a robot at position (0, 0). The robot can only move diagonally, once place at a time,

Reviewer 5: "once place at a time" => "one place at a time"
(reviewing commit by commit annoyingly doesn't let me use "suggestions"...)

Reviewer 5: Maybe actually "one step at a time" :-).

            require(y > type(int128).min && y < type(int128).max);
            _;
        }

        function moveLeftUp() wall public {
            --x;
            ++y;
        }

        function moveLeftDown() wall public {
            --x;
            --y;
        }

        function moveRightUp() wall public {
            ++x;
            ++y;
        }

        function moveRightDown() wall public {
            ++x;
            --y;
        }

        function inv() public view {
            assert((x + y) % 2 == 0);
        }
    }

Function ``inv`` represents an invariant of the state machine that ``x + y``
must be even.
The SMTChecker manages to prove that regardless how many commands we give the
robot, even if infinitely many, the invariant can *never* fail. The interested
reader may want to prove that fact manually as well.  Hint: this invariant is
inductive.

We can also trick the SMTChecker into giving us a path to a certain position.
We know that position (2, 4) is reachable, because 2 + 4 = 6 is even. Let us add
the property that (2, 4) is *not* reachable, by adding the following function.

.. code-block:: Solidity

    function reach_2_4() public view {

>         assert(!(x == 2 && y == 4));

Reviewer 4: I was curious if it would try to produce the transaction trace when the numbers `2` and `4` are replaced `2**120`.
Apparently it gives just

```

Warning: CHC: Error trying to invoke SMT solver.
  --> /tmp/smt.sol:40:13:
   |
40 |             assert(!(x == 2**120 && y == 2**120));
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Warning: CHC: Assertion violation might happen here.
  --> /tmp/smt.sol:40:13:
   |
40 |             assert(!(x == 2**120 && y == 2**120));
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Warning: BMC: Assertion violation happens here.
  --> /tmp/smt.sol:40:13:
   |
40 |             assert(!(x == 2**120 && y == 2**120));
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: Counterexample:
  x = 2**120
  y = 2**120

Note: Callstack:
Note:

```
The empty `note` looks weird.

Reviewer 4: It's a nice example to illustrate CHC, though!

Author: The problem with using 2^120 is that it then needs to find a counterexample of size 2**120, since that's the only way to break the assertion, that's why it fails.

Re empty note: true, I ran everything with CHC only.

Author: It should ofc not error out, that's probably an issue in Spacer's cex generation. But I'm fairly certain it wouldn't solve this, the cex is just too large (2^120 transactions)

    }

We know that this property is false, but while proving that the property is false,

> the SMTChecker gives us exactly *how* to reach (2, 4):

Reviewer 5: "gives" -> "tells"?

available at compile time, there is no guarantee that the deployed contract
will indeed be the same as the contract where the interface came from at
compile time.

In some cases, it is possible to automatically infer properties over state
variables that are still true even if the externally called code can do
anything, including reenter the caller contract.

.. code-block:: Solidity

  pragma experimental SMTChecker;
  
  interface Unknown {
  	function run() external;
  }
  
  contract Mutex {
  	uint x;
  	bool lock;
  
  	Unknown immutable unknown;
  
  	constructor(Unknown _u) {
  		require(address(_u) != address(0));
  		unknown = _u;
  	}
  
  	modifier mutex {
  		require(!lock);
  		lock = true;
  		_;
  		lock = false;
  	}
  
  	function set(uint _x) mutex public {
  		x = _x;
  	}
  
  	function run() mutex public {
  		uint xPre = x;
  		unknown.run();
  		assert(xPre == x);
  	}
  }

The example above shows a contract that uses a mutex flag to forbid reentrancy.
The solver is able to infer that when ``unknown.run()`` is called, the contract
is already "locked", so it would not be possible to change the value of ``x``,

> regardless what the unknown called code does.

Reviewer 5: "regardless of what..."? Or "no matter what"?

     | 		^^^^^^^^^^^^^^^^^

******************
SMTChecker Options
******************

Model Checking Engines
======================

The SMTChecker module implements two different reasoning engines, a Bounded
Model Checker (BMC) and a system of Constrained Horn Clauses (CHC).  Both
engines are currently under development, and have different characteristics.
The engines are independent and every property warning states from which engine

> it came from. Notice that all the examples above with counterexamples were

Reviewer 5: Loose the "from" at the end. Or "and every property warning states which engine issued it."
Also "Note that" instead of "Notice that" maybe?

reported by CHC, the more powerful engine.

By default both engines are used, where CHC runs first, and every property that
was not proven is passed over to BMC. You can choose a specific engine via the CLI
option ``--model-checker-engine {all,bmc,chc,none}`` or the JSON option
``settings.modelChecker.engine={all,bmc,chc,none}``.

Bounded Model Checker (BMC)
---------------------------

The BMC engine analyzes functions in isolation, that is, it does not take the

> overall behavior of the contract throughout many transactions into account when

Reviewer 5: "throughout many" -> "over multiple"?
Two are enough, aren't they?

analyzing each function.  Loops are also ignored in this engine at the moment.

> Internal function calls are inlined as long as they are not recursive, direct

Reviewer 5: direct*ly*
I'd also make it "as long as they are not (directly or indirectly) recursive."

Author: I think you can omit the suffix here if the next word has it, here "direct or indirectly"

Reviewer 5: Never heard that one before :-D...

Reviewer 5: The grammar processor in my brain highly disagrees in any case :-).

Reviewer 5: `or :: 'a => 'a => 'a`, `direct :: adjective`, `indirectly :: adverb`,
`(or) direct indirectly` - clash of types :-).

Author: What?
Here's an example: https://ieeexplore.ieee.org/document/8097218

Reviewer 5: To me it would seem the example either reads "(direct) and (indirectly liquid-cooled) coreless linear actuators" or (maybe in that particular case more likely) is also just wrong :-).

Reviewer 5: But yeah for the record: fine with keeping it, even though I think it's wrong :-). Maybe I'm wrong about it being wrong :-).

Author: I haven't found proof, and your case seems more consistent so I'll change it :)

> or indirectly. External function calls are inlined if possible, and knowledge

Reviewer 5: The knowledge erasure is independent of the inlining isn't it? I'd make it
"External function calls are inlined, if possible. Knowledge..."

Maybe I'm getting too nitpickey, though :-D.

Author: Correct

that is potentially affected by reentrancy is erased.

> The characteristics above make BMC easily prone to reporting false positives,

Reviewer 5: "prone to" is enough without "easily" I'd say.

but it is also lightweight and should be able to quickly find small local bugs.

Constrained Horn Clauses (CHC)
------------------------------

> The Solidity contract's Control Flow Graph (CFG) is modelled as a system of

Reviewer 5: "The contract's Control Flow Graph (CFG)" or maybe even rather "The Control Flow Graph (CFG) of a contract". No need to state that it's Solidity again, is there?

> Horn clauses, where the life cycle of the contract is represented by a loop

Reviewer 5: "loop" -> "cycle"? Not sure though, fine either way...
EDIT: but maybe really better, since you talk about "for-loop"-kind-loops below as well...

Author: also not to repeat "life cycle ... cycle"

Reviewer 5: Yeah, ok, just found it a bit confusing to talk about loops in two different meanings like that there, but maybe it's fine!
